### PR TITLE
Classes

### PR DESCRIPTION
**Pull request by timmooore**
Merge Sprint 4 new features into main. Requires code review.

### CODE REVIEW

File: src/models/class.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const mongoose = require("mongoose");
+const Schema = mongoose.Schema;
+const passportLocalMongoose = require("passport-local-mongoose");
```
Comment (AleCondina) [inline src/models/class.js]: The `passport-local-mongoose` module is imported but not used in this file. If it’s not required, consider removing it to clean up unused dependencies. If it will be used later, please add a comment explaining its intended purpose.

File: src/models/class.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+const ClassSchema = new Schema({
```
Comment (AleCondina) [inline src/models/class.js:4]: Good use of `required: true` to enforce that every class has a name. Consider adding additional validation if needed (ensuring unique class names, if applicable).

File: src/models/class.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  class_name: {
+    type: String,
+    required: true,
+  },
+  instructor_id: {
```
Comment (AleCondina) [inline src/models/class.js:9]: The reference to the `User` schema looks appropriate. Ensure validation logic is added to verify that the referenced instructor actually exists before saving a `Class` object.

File: src/models/class.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    type: Schema.Types.ObjectId,
+    ref: "User", // Reference to the instructor teaching the class
+    required: true,
+  },
+  student_ids: [
```
Comment (AleCondina) [inline src/models/class.js:14]: The array structure for `student_ids` looks good. To improve robustness, consider adding middleware or validation to prevent duplicate student entries in this array.

File: src/models/class.js
Comment (AleCondina) [inline src/models/class.js:1]: -Consider adding hooks (example: `pre-save` hooks) to enforce additional validation rules or ensure logical consistency in the data being saved to the database. 

-Consider indexing frequently queried fields such as `instructor_id` and `student_ids` to improve database performance for common operations.

Please ensure that this schema is covered with appropriate tests. Key cases include:
- Ensuring `class_name` and `instructor_id` are required.
- Validating the addition/removal of `student_ids`.
- Verifying relationships with the `User` schema (e.g., ensuring `populate()` works as expected).

File: src/models/user.js [hunk 5]
```diff
@@ -5,14 +5,17 @@ const passportLocalMongoose = require("passport-local-mongoose");
 const UserSchema = new Schema({
   user_type: {
     type: String,
-    enum: ["student", "instructor"], //Only allows users to be "students" or "instructors"
+    enum: ["student", "instructor"], // Only allows users to be "students" or "instructors"
     required: true,
   },
+  classes: [
```
Comment (AleCondina) [inline src/models/user.js:11]: To improve robustness:
- Ensure validation checks are added to confirm that all referenced `Class` documents exist.
- Consider adding a mechanism to prevent duplicate class entries in this array.

File: src/models/user.js
Comment (AleCondina) [inline src/models/user.js:1]: Consider adding back a short explanation for future contributors, especially since it handles username/password logic implicitly.

File: src/routes/assessmentRoutes.js [hunk 6]
```diff
@@ -2,37 +2,72 @@ const express = require("express");
 const { isLoggedIn, isInstructor } = require("../middlewares/middleware");
```
Comment (AleCondina) [inline src/routes/assessmentRoutes.js:2]: Consider adding specific error messages or redirects in the middleware itself for consistency.

File: src/routes/assessmentRoutes.js [hunk 7]
```diff
@@ -2,37 +2,72 @@ const express = require("express");
...
 const User = require("../models/user");
 const Evaluation = require("../models/evaluation");
+const Class = require("../models/class");
 const router = express.Router();
 
 //Instructor Assessment View Page
```
Comment (AleCondina) [inline src/routes/assessmentRoutes.js:8]: Add validation middleware for classId

File: src/routes/assessmentRoutes.js [hunk 8]
```diff
@@ -71,84 +106,139 @@ router.get("/assessments/:studentId", isLoggedIn, async (req, res) => {
...
+      const studentIds = [...new Set(evaluations.map((eval) => eval.evaluatee))];
 
-      // Fetch evaluations for each student without populating evaluator
+      // Fetch student details for those evaluatees
+      const students = await User.find({
+        _id: { $in: studentIds },
+        user_type: "student",
+      });
+
+      // Fetch evaluations for each student
```
Comment (AleCondina) [inline src/routes/assessmentRoutes.js:156]: Add a check to ensure students is not empty before proceeding to fetch evaluations.

File: src/routes/assessmentRoutes.js [hunk 9]
```diff
@@ -71,84 +106,139 @@ router.get("/assessments/:studentId", isLoggedIn, async (req, res) => {
...
+      const evaluations = await Evaluation.find({ class_id: classId });
+      const studentIds = [...new Set(evaluations.map((eval) => eval.evaluatee))];
+
+      // Fetch detailed student records
+      const students = await User.find({
+        _id: { $in: studentIds },
+        user_type: "student",
+      });
+
+      // Aggregate average ratings and scores for each student
```
Comment (AleCondina) [inline src/routes/assessmentRoutes.js:209]: Handle cases where totals.count is 0 to avoid potential division by zero errors.

File: src/routes/assessmentRoutes.js
Comment (AleCondina) [inline src/routes/assessmentRoutes.js:1]: code is well written overall. Comments for code blocks explaining why they are there should be clarified.

File: src/routes/teamsRoutes.js [hunk 10]
```diff
@@ -1,79 +1,164 @@
 const express = require("express");
-const { isLoggedIn, isInstructor } = require("../middlewares/middleware");
+const { isLoggedIn, isInstructor, isStudent } = require("../middlewares/middleware");
```
Comment (AleCondina) [inline src/routes/teamsRoutes.js]: The `isStudent` middleware is imported but never used. If it's unnecessary, remove it to clean up unused dependencies and avoid linting errors. If it will be used in future routes, include a comment explaining its purpose.

File: src/routes/teamsRoutes.js
Comment (AleCondina) [inline src/routes/teamsRoutes.js:1]: Consider Refactoring repeated validation logic into middleware, improving error logging, and optimizing database queries. It will make the codebase more maintainable and robust.

File: src/routes/authRoutes.js [hunk 11]
```diff
@@ -3,45 +3,61 @@ const passport = require("passport");
 const User = require("../models/user");
 const router = express.Router();
 
+// Render the registration page
 router.get("/register", (req, res) => res.render("register"));
+
+// Handle registration logic
 router.post("/register", async (req, res, next) => {
   try {
     const { username, password, user_type } = req.body;
```
Comment (AleCondina) [inline src/routes/authRoutes.js:12]: Add validation logic for user_type or use a middleware for this.

File: src/routes/authRoutes.js [hunk 12]
```diff
@@ -3,45 +3,61 @@ const passport = require("passport");
...
     const user = new User({ username, user_type });
     const registeredUser = await User.register(user, password);
     req.login(registeredUser, (err) => {
-      if (err) return next(err); //Callback function if an error occurs
-      // If they're an instructor, go to the instructor dashboard; otherwise, go to the student
-      // dashboard
+      if (err) {
```
Comment (AleCondina) [inline src/routes/authRoutes.js:16]: Log the error details for development purposes.

File: src/routes/authRoutes.js [hunk 13]
```diff
@@ -3,45 +3,61 @@ const passport = require("passport");
...
     });
-  } catch {
-    res.redirect("register"); //Something hrouterened, so go back to the register page
+  } catch (err) {
+    req.flash("error", "Registration failed. Username might already be taken.");
+    res.redirect("/register");
   }
 });
 
+// Render the login page
```
Comment (AleCondina) [inline src/routes/authRoutes.js:39]: Remove the err parameter if it’s not used in the logout route

File: src/routes/authRoutes.js
Comment (AleCondina) [inline src/routes/authRoutes.js:1]: Resolve the linting issue by removing the unused `err` variable in the logout route or by handling it explicitly.

Write tests for the following:
- Successful and failed registration attempts (e.g., duplicate username).
- Login attempts with valid and invalid credentials.
- Logout behavior and redirection.

File: src/routes/instructorRoutes.js [hunk 14]
```diff
@@ -1,18 +1,44 @@
 const express = require("express");
 const { isLoggedIn, isInstructor } = require("../middlewares/middleware");
 const Team = require("../models/team");
+const Class = require('../models/class');
```
Comment (AleCondina) [inline src/routes/instructorRoutes.js:4]: Add middleware for validating input fields (example: class_name) to ensure data integrity during requests.

File: src/routes/instructorRoutes.js
Comment (AleCondina) [inline src/routes/instructorRoutes.js:1]: Write tests for the following:
- Successful class creation.
- Attempting to create a class without `class_name`.
- Fetching classes for an instructor with and without any associated classes.
- Accessing routes without the proper role (e.g., as a student or unauthenticated user).

File: src/routes/miscRoutes.js
Comment (AleCondina) [inline src/routes/miscRoutes.js:1]: Log additional contextual information (example: `req.user._id`, `classId`) in error logs to simplify debugging.

For team creation and user assignment:
- Use batch operations (example: `insertMany`) to reduce the number of individual database writes.
- Add indexing for frequently queried fields like `username` and `class_id` to improve query performance.

Write tests to cover:
- Valid and invalid CSV uploads.
- Team creation with various team sizes.
- Password update flow, including invalid current passwords and mismatched new passwords.

File: src/routes/studentRoutes.js
Comment (AleCondina) [inline src/routes/studentRoutes.js:1]: Write tests to cover:
- Access control (e.g., ensuring only students can access these routes).
- Valid and invalid inputs (e.g., invalid `classId`, missing parameters).
- Assessment submission with valid and invalid ratings.
- Edge cases (e.g., a student without any teams or classes).

**Review by AleCondina — COMMENTED** [review]
General Suggestions for Improvement: 

-implement input validation middleware  for all routes 
 
-Sanitize user inputs to prevent injection attacks and replace default passwords with secure credential management methods. 

-Improve error feedback by logging contextual details  and providing user friendly messages. 

-Optimize database queries by projecting necessary fields, indexing frequently queried ones, and adding pagination for large datasets. 

-Extract repeated logic, like class validation, into reusable middleware to improve maintainability. 

-Enhance templates to handle empty states gracefully and add client side validation for forms to improve user experience.

-Write comprehensive tests to cover edge cases, access control, and invalid inputs for a robust application. (see comments for suggested tests)

**Comment by AleCondina** [comment]
I ran the branch and did some general UX testing. Everything is working well from the user side.

File: src/models/class.js (hunk 1 above)
Comment (timmooore) [reply src/models/class.js]: Module was moved and not needed here.
